Set appropriate `stacklevel=` in all calls to `warnings.warn()`

I noticed [in this lovely video](https://youtu.be/mkgd9iOiICc?t=1039) that the warning for `st.integers().example()` pointed into the Hypothesis internals, rather than at the calling user code.  The solution is to set `stacklevel=` in:

- [x] `SearchStrategy.example()` in `strategies.py`
- [x] `warn_on_missing_dtypes()` in `array_api.py` - this might need to accept a new `stacklevel=` argument, since it's called from multiple different places.  [turns out not to be worth the trouble 😞]

Test plan: manually check that these warnings now point to user code - the problem isn't that we break such handling but rather that we hadn't added it in the first place.

I've created the following minimal, reproducible example for this.

<details><summary>summary of versions (click me)</summary>

```text
platform darwin -- Python 3.9.13, pytest-7.1.2, pluggy-1.0.0
rootdir: /Users/jlamb/repos/hypothesis, configfile: pytest.ini
plugins: xdist-2.5.0, forked-1.4.0, hypothesis-6.50.1
```

</details>

Given a file `test_warnings.py` with the following.

```python
from hypothesis import (
    example,
    given,
    strategies as st
)


def _bad_max(x, y):
    if x == 2:
        return -1
    else:
        return max(x, y)


@given(st.integers(), st.integers())
@example(2, st.integers().example())
def test_max_works(x, y):
    assert _bad_max(x, y) == max(x, y)
```

Running this with all of `pytest`'s defaults

```shell
pytest test_warnings.py
```

Returns a non-0 exit code as expected, and produces output that shows user code.

```text
================================================= FAILURES ==================================================
______________________________________________ test_max_works _______________________________________________
Traceback (most recent call last):
  File "/Users/jlamb/repos/hypothesis/test_warnings.py", line 16, in test_max_works
    @example(2, st.integers().example())
  File "/Users/jlamb/repos/hypothesis/hypothesis-python/src/hypothesis/core.py", line 1208, in wrapped_test
    raise the_error_hypothesis_found
  File "/Users/jlamb/repos/hypothesis/test_warnings.py", line 18, in test_max_works
    assert _bad_max(x, y) == max(x, y)
AssertionError: assert -1 == 3263640430859197895
 +  where -1 = _bad_max(2, 3263640430859197895)
 +  and   3263640430859197895 = max(2, 3263640430859197895)
------------------------------------------------ Hypothesis -------------------------------------------------
Falsifying explicit example: test_max_works(
    x=2, y=3263640430859197895,
)
```

But using warnings-as-errors, Hypothesis's warning is printed instead.

```shell
pytest -Werror ./test_warnings.py
```

Returns a non-0 exit code, but doesn't show user code.

```text
================================================== ERRORS ===================================================
_____________________________________ ERROR collecting test_warnings.py _____________________________________
Traceback (most recent call last):
  File "/Users/jlamb/repos/hypothesis/test_warnings.py", line 16, in <module>
    @example(2, st.integers().example())
  File "/Users/jlamb/repos/hypothesis/hypothesis-python/src/hypothesis/strategies/_internal/strategies.py", line 285, in example
    warnings.warn(
hypothesis.errors.NonInteractiveExampleWarning: The `.example()` method is good for exploring strategies, but should only be used interactively.  We recommend using `@given` for tests - it performs better, saves and replays failures to avoid flakiness, and reports minimal examples. (strategy: integers())
========================================== short test summary info ==========================================
ERROR test_warnings.py - hypothesis.errors.NonInteractiveExampleWarning: The `.example()` method is good f...
```

Cool, cool ok.

So I think that both in my example (error at test collection time) and your example (error raised directly in the test codde), `hypothesis's` behavior is what I'd expect.

You get a line in the traceback pointing to the specific test file + line in that test file which caused the warning.

Running your example with `pytest -Werror test_warnings.py`, I see this:

```text
================================================= FAILURES ==================================================
___________________________________________________ test ____________________________________________________
Traceback (most recent call last):
  File "/Users/jlamb/repos/hypothesis/test_warnings.py", line 24, in test
    st.integers().example()
  File "/Users/jlamb/repos/hypothesis/hypothesis-python/src/hypothesis/strategies/_internal/strategies.py", line 286, in example
    warnings.warn(
hypothesis.errors.NonInteractiveExampleWarning: The `.example()` method is good for exploring strategies...
```

That seems pretty good to me. Can you help me understand what type of output you'd expect to see instead? I might be misunderstanding what "point to user code" means.

--- a/AUTHORS.rst
+++ b/AUTHORS.rst
@@ -73,6 +73,7 @@ their individual contributions.
 * `Ivan Tham <https://github.com/pickfire>`_
 * `Jack Massey <https://github.com/massey101>`_
 * `Jakub Nabaglo <https://github.com/nbgl>`_ (j@nab.gl)
+* `James Lamb <https://github.com/jameslamb>`_
 * `Jenny Rouleau <https://github.com/jennyrou>`_
 * `Jeremy Thurgood <https://github.com/jerith>`_
 * `J.J. Green <http://soliton.vm.bytemark.co.uk/pub/jjg/>`_

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,6 @@
+RELEASE_TYPE: patch
+
+This patch ensures that the warning for non-interactive ``.example()``
+points to your code instead of Hypothesis internals (:issue:`3403`).
+
+Thanks to @jameslamb for this fix.

--- a/hypothesis-python/src/hypothesis/strategies/_internal/strategies.py
+++ b/hypothesis-python/src/hypothesis/strategies/_internal/strategies.py
@@ -288,6 +288,7 @@ class SearchStrategy(Generic[Ex]):
                 "it performs better, saves and replays failures to avoid flakiness, "
                 "and reports minimal examples. (strategy: %r)" % (self,),
                 NonInteractiveExampleWarning,
+                stacklevel=2,
             )
 
         context = _current_build_context.value

